src/native/eventpipe/ep-buffer.c

```diff
@@ -196,11 +250,63 @@ ep_buffer_convert_to_read_only (EventPipeBuffer *buffer)
 	ep_rt_volatile_store_uint32_t (&buffer->state, (uint32_t)EP_BUFFER_STATE_READ_ONLY);
 
 	// If this buffer contains an event, select it.
-	uint8_t *first_aligned_instance = ep_buffer_get_next_aligned_address (buffer, buffer->buffer);
+	uint8_t *first_aligned_instance = buffer->first_event_address;
 	if (buffer->current > first_aligned_instance)
 		buffer->current_read_event = (EventPipeEventInstance*)first_aligned_instance;
 	else
 		buffer->current_read_event = NULL;
+
+	if (buffer->buffer_guard_level != EP_BUFFER_GUARD_LEVEL_NONE)
+		ep_buffer_ensure_guard_consistency (buffer);
+
+	if ((buffer->buffer_guard_level >= EP_BUFFER_GUARD_LEVEL_PROTECT_ON_READONLY) &&
```

Comment:
same thing, this if can be nested under the one at line 259 to ensure the compiler can keep it off the hot path.